### PR TITLE
Client 6885 | ICE Gathering failures and timeouts issue ICE Restarts

### PR DESCRIPTION
**Pull request by charliesantos**
[CLIENT-6885](https://issues.corp.twilio.com/browse/CLIENT-6885) ICE Gathering failures and timeouts issue ICE Restarts.

### CODE REVIEW

File: lib/twilio/rtc/peerconnection.js [hunk 1]
```diff
@@ -654,11 +690,38 @@ PeerConnection.prototype._setupChannel = function() {
...
+      this._stopIceGatheringTimeout();
+
+      // Fail if no candidates found
+      if (!this._hasIceCandidates) {
+        this._onIceGatheringFailure(ICE_GATHERING_FAIL_TIMEOUT);
+      }
+
+      // There was a failure mid-gathering phase. We want to start our timer and issue
+      // an ice restart if we don't get connected after our timeout
+      if (this._hasIceCandidates && this._hasIceGatheringFailures) {
```
Comment (charliesantos) [inline lib/twilio/rtc/peerconnection.js]: I'm not a fan of using more states in rtc/peerconnection, but I'm not sure how else we can capture these values without using boolean states. I'm open to suggestions!

File: lib/twilio/connection.ts [hunk 2]
```diff
@@ -337,6 +337,11 @@ class Connection extends EventEmitter {
       this._publisher.post(level, 'ice-connection-state', state, null, this);
     };
 
+    this.mediaStream.onicegatheringfailure = (type: string): void => {
+      this._publisher.warn('ice-gathering-state', type, null, this);
```
Comment (ryan-rowland) [inline lib/twilio/connection.ts]: What can this be, other than 'failed'?

Comment (ryan-rowland) [reply lib/twilio/connection.ts]: Does this map to the following?
```
const ICE_GATHERING_FAIL_NONE = 'none';
const ICE_GATHERING_FAIL_TIMEOUT = 'timeout';
```

We should enumerate this, it's not clear what "type" can be.

Also, it looks like there's a case mistmatch here in `icegatheringfailure`

Comment (charliesantos) [reply lib/twilio/connection.ts]: Ok, I will add an enum for 'none' and 'timeout'.

Are you referring to `_onIceGatheringFailure` and `onicegatheringfailure`

Comment (ryan-rowland) [reply lib/twilio/connection.ts]: Nevermind on `onicegatheringfailure`, I see.

Comment (charliesantos) [reply lib/twilio/connection.ts]: `_onIceGatheringFailure ` is a private method which calls `onicegatheringfailure ` which is a callback.  https://github.com/twilio/twilio-client.js/pull/94/files#diff-d6d1e696fa57bedaec729f3d6c8274a1R381

**Pull request merged**